src/datamodel/builder/Uuid.cpp

```diff
@@ -10,3 +10,99 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include "up-cpp/datamodel/builder/Uuid.h"
+
+#include <random>
+#include <stdexcept>
+
+namespace uprotocol::datamodel::builder {
+
+struct UuidBuilder::UuidSharedState {
+	std::mt19937_64 random_engine{std::random_device{}()};
+	uint16_t counter{0};
+	uint64_t last_unix_ts_ms{0};
+	uint64_t rand_b{0};
+	bool rand_b_initialized{false};
+};
+
+UuidBuilder UuidBuilder::getBuilder() { return UuidBuilder(false); }
+
+UuidBuilder UuidBuilder::getTestBuilder() { return UuidBuilder(true); }
+
+UuidBuilder& UuidBuilder::withTimeSource(
+    std::function<std::chrono::system_clock::time_point()>&& time_source) {
+	if (!testing_) {
+		throw std::domain_error(
+		    "Cannot set time source on non-test UuidBuilder");
+	}
+	time_source_ = std::move(time_source);
+	return *this;
+}
+
+UuidBuilder& UuidBuilder::withRandomSource(
+    std::function<uint64_t()>&& random_source) {
+	if (!testing_) {
+		throw std::domain_error(
+		    "Cannot set random source on non-test UuidBuilder");
+	}
+	random_source_ = std::move(random_source);
+	return *this;
+}
+
+UuidBuilder& UuidBuilder::withIndependentState() {
+	if (!testing_) {
+		throw std::domain_error(
+		    "Cannot set independent state on non-test UuidBuilder");
+	}
+	shared_state_ = std::make_shared<UuidSharedState>();
+	return *this;
+}
+
+v1::UUID UuidBuilder::build() {
+	v1::UUID uuid;
+	auto now = time_source_ ? time_source_() : std::chrono::system_clock::now();
+	auto unix_ts_ms = std::chrono::duration_cast<std::chrono::milliseconds>(
+	                      now.time_since_epoch())
+	                      .count();
+
+	if (unix_ts_ms != shared_state_->last_unix_ts_ms) {
+		// Reset the counter if the timestamp tick has changed
+		shared_state_->counter = 0;
+		shared_state_->last_unix_ts_ms = unix_ts_ms;
+	}
+
+	uint64_t msb = static_cast<uint64_t>(unix_ts_ms) << 16;
```

Comment:
This looks more like UUIDv7 instead of UUIDv8 (v7 says time, v8 says implementation specific).

Reply:
Agree, the uuidv8 spec we have is very similar to v7 in term of bits arrangement

Reply:
I'll follow up with another PR to make this adhere with current v7 spec